src/charm.py

```diff
@@ -63,14 +66,24 @@ def __init__(self, *args):
         self.service_patch = KubernetesServicePatch(self, [(self.app.name, self._port)])
 
         self.alertmanager_consumer = AlertmanagerConsumer(self, relation_name="alertmanager")
+        self.ingress_per_unit = IngressPerUnitRequirer(
+            self, relation_name="ingress", port=self._port
+        )
+        external_url = urlparse(self._external_url)
```

Comment:
This looks a little confusing, mostly because `_external_url` is defined as a property waaaaay at the bottom, and it's not obvious when reading the constructor.

Especially considering that the parts aren't used until a little below this, it may read better by moving it down to just above `LokiPushApiProvider` and renaming the variable to something which can be distinguished at first-pass (instead of `self._external_url <-> external_url`), like `parsed_external_url` or whatever name you like.

Reply:
Done